### PR TITLE
Move metrics to context

### PR DESCRIPTION
**Pull request by mvpatel2000**

### CODE REVIEW

**Review by Landanjs — APPROVED** [review]
LGTM

File: composer/trainer/trainer.py
Comment (dskhudia) [inline composer/trainer/trainer.py:1503]: You can just do `torch.autocast(enabled=False)` to make it device and dtype agnostic.

Comment (mvpatel2000) [reply composer/trainer/trainer.py]: ([JIRA](https://mosaicml.atlassian.net/browse/CO-827)) so the issue is if training happens in lower precision, metrics might blow up if not computed in same precision. Tried always setting it to float 32, but realizing actually there might be stuff that would require lower floats, so I'm moving this to same precision context as `validate()`

Comment (abhi-mosaic) [reply composer/trainer/trainer.py]: Are there really metrics that would be computed in low precision? I feel like it kinda makes sense to always do FP32... I even see a comment in the Trainer that we revert metrics to FP32 when DeepSpeed accidentally downcasts them.

Also @dskhudia I think that `torch.autocast(enabled=False)` would be like a no-op, but what we want is to force the inputs to metrics like softmax to be upcast to `torch.float32` (they are coming in as 16 bit), so I think the dtype might be necessary?

Comment (abhi-mosaic) [reply composer/trainer/trainer.py]: Alternatively, if we have some general function that can just convert a batch to a dtype, then all we really need to do is call `upcast(outputs)` and `upcast(targets)`

Comment (dskhudia) [reply composer/trainer/trainer.py]: `torch.autocast(enabled=False)` will disable autocasting but inputs under this context should be upcast. So your `upcast(outputs)` makes sense and a reasonable way to do it.

Comment (dskhudia) [reply composer/trainer/trainer.py]: Also agree with doing metrics update with fp32.

Comment (mvpatel2000) [reply composer/trainer/trainer.py]: TLDR: It looks like its pretty hard to autocast / upcast to fp32 on CPU. You have to manually cast each tensor, but model output can be really messy. Torch handles a lot of stuff internally well (eg Abhi found + casts to fp32), so after offline discussion, i think we should leave PR as is

Comment (mvpatel2000) [reply composer/trainer/trainer.py]: @dskhudia if this sounds reasonable to you, can you approve + resolve comments? need approval from code owner

Comment (dskhudia) [reply composer/trainer/trainer.py]: sounds good.

**Review by dskhudia — APPROVED** [review]

**Pull request merged**
